src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -5665,9 +5685,65 @@ bool StorageReplicatedMergeTree::waitForShrinkingQueueSize(size_t queue_size, UI
     return true;
 }
 
+bool StorageReplicatedMergeTree::dropPart(
+    zkutil::ZooKeeperPtr & zookeeper, String part_name, LogEntry & entry, bool detach)
+{
+    LOG_TRACE(log, "Will try to insert a log entry to DROP_RANGE for part: " + part_name);
+
+    auto part_info = MergeTreePartInfo::fromPartName(part_name, format_version);
+
+    while (true)
+    {
+        ReplicatedMergeTreeMergePredicate merge_pred = queue.getMergePredicate(zookeeper);
+
+        /// TODO(nv) It is possible that part does not exist on replica which executes this statement.
+        ///     Also, it possible for the part to not exist on any replicas, replica which created log entries for the part disappeared.
+        auto part = data_parts_by_info.find(part_info);
+        if (part == data_parts_by_info.end())
+            throw Exception("Part " + part_name + " not found locally, won't try to drop it.", ErrorCodes::NOT_IMPLEMENTED);
+
+        /// TODO(nv) get ops and commit together w/ log entry?
```

Comment:
How principled are you about this one, @alexey-milovidov ?

Reply:
Yes, it must be done in a single transaction